### PR TITLE
Add configurable query options: timeout, IP TTL, NTP Server port

### PR DESCRIPTION
**Pull request by darkk**
I need SetTTL(1) to limit my code talking to NTP servers within LAN as a precaution measure (to avoid overloading NTP pool in case of misconfiguration), so I convert bare _timeout_ to structure of options and add it there.

### CODE REVIEW

File: ntp.go [hunk 1]
```diff
@@ -139,12 +144,23 @@ type Response struct {
 	Leap           LeapIndicator // server's leap second indicator; see RFC 5905
 }
 
+type QOption struct {
```
Comment (beevik) [inline ntp.go]: Let's rename this QueryOptions. And add a comment that will be used for the documentation.  Something like:

```go
// QueryOptions contains the list of configurable options that may be used with the Query function.
```

File: ntp.go [hunk 2]
```diff
@@ -139,12 +144,23 @@ type Response struct {
...
+	Timeout time.Duration // defaults to defaultTimeout
```
Comment (beevik) [inline ntp.go]: This comment should be

`// defaults to 5 seconds`

since `defaultTimeout` is private and will not be visible in the documentation.

(I think it's fine and appropriate that defaultTimeout is private, by the way.)

File: ntp.go [hunk 3]
```diff
@@ -139,12 +144,23 @@ type Response struct {
...
+	Version int           // NTP protocol version, defaults to 4
+	Port    int           // NTP Server port for UDPAddr.Port, defaults to 123
+	IpTTL   int           // IP TTL to use for outgoing UDP packets
```
Comment (beevik) [inline ntp.go]: idiomatic go does not "camel case" acronyms.  So this should either be `IPTTL` or just `TTL`.  I prefer `TTL`.  The comment clarifies that it is for the IP protocol.

File: ntp.go [hunk 4]
```diff
@@ -139,12 +144,23 @@ type Response struct {
...
+}
+
 // Query returns the current time from the remote server host using the
 // requested version of the NTP protocol. It also returns additional
 // information about the exchanged time information. The version may be 2, 3,
 // or 4; although 4 is most typically used.
 func Query(host string, version int) (*Response, error) {
-	m, err := getTime(host, version)
+	return QueryEx(host, QOption{Version: version})
```
Comment (beevik) [inline ntp.go]: Let's change the function `Query` to simply take the `host` string, and remove the `version`.  Then your new function can be the one that takes all the options.  The comment will need revision.

(This is technically an API breaking change, but a minor one, and `Query` is somewhat new to this package anyway.)

File: ntp.go [hunk 5]
```diff
@@ -139,12 +144,23 @@ type Response struct {
...
+}
+
+func QueryEx(host string, opt QOption) (*Response, error) {
```
Comment (beevik) [inline ntp.go]: Let's call this `QueryWithOptions` instead.  I'm not a big fan of `Ex` APIs when they can be avoided.  This function will also need a comment for the documentation.

File: ntp.go [hunk 6]
```diff
@@ -173,27 +189,49 @@ func Query(host string, version int) (*Response, error) {
...
+
 	con, err := net.DialUDP("udp", nil, raddr)
 	if err != nil {
 		return nil, err
 	}
 	defer con.Close()
-	con.SetDeadline(time.Now().Add(timeout))
+
+	if opt.IpTTL != 0 {
+		ipcon := ipv4.NewConn(con)
```
Comment (beevik) [inline ntp.go]: I'm not familiar with the `golang.org/x/net/ipv4` package.  Is this all you have to do to set a TTL on packets sent through the `con`?  Are you sure you don't have to declare the `ipcon` outside the `if` statement and use it instead of `con` in the `binary.Write` call below?

File: ntp.go [hunk 7]
```diff
@@ -173,27 +189,49 @@ func Query(host string, version int) (*Response, error) {
...
+	}
+
+	con.SetDeadline(time.Now().Add(opt.Timeout))
 
 	m := new(msg)
 	m.setMode(client)
-	m.setVersion(version)
-	m.TransmitTime = toNtpTime(time.Now())
+	m.setVersion(opt.Version)
+	xmt := toNtpTime(time.Now())
```
Comment (beevik) [inline ntp.go]: Why did you need to introduce the `xmt` temporary variable?  Can't you simply compare `m.OriginTime` to `m.TransmitTime` below?

File: ntp.go [hunk 8]
```diff
@@ -205,14 +243,30 @@ func getTime(host string, version int) (*msg, error) {
...
 
+	// It's possible to use random uint64 as client's `TransmitTime` field,
+	// it has better privacy (clock of the node is not disclosed in
+	// plain-text), better UDP packet spoofing resistance (blind attacker
+	// has to guess both port and the uint64 value), and OpenNTPD behaves
+	// like that. But math/rand is not secure enough for the purpose,
+	// crypto/rand takes 64 bits of entropy for every outgoing packet and
+	// CSPRNG from crypto/rand/rand_unix is not available: see
+	// https://github.com/golang/go/issues/13820
+	if m.OriginTime != xmt {
```
Comment (beevik) [inline ntp.go]: Do you have a reference to the section of the RFC that mandates the origin time should be equal to the transmit time?

File: ntp.go [hunk 9]
```diff
@@ -205,14 +243,30 @@ func getTime(host string, version int) (*msg, error) {
...
 
 // TimeV returns the current time from the remote server host using the
 // requested version of the NTP protocol. The version may be 2, 3, or 4;
 // although 4 is most typically used.
 func TimeV(host string, version int) (time.Time, error) {
-	m, err := getTime(host, version)
+	return TimeEx(host, QOption{Version: version})
+}
+
+func TimeEx(host string, opt QOption) (time.Time, error) {
```
Comment (beevik) [inline ntp.go]: The `QueryWithOptions` function is the one people can use if they care about more advanced options.  I'd like to keep the `Time` functions as simple as possible.  So let's not use `TimeEx`.

File: ntp.go (hunk 4 above)
Comment (darkk) [inline ntp.go]: > API breaking change

Yep, that's exactly what I was trying to avoid :)

File: ntp.go (hunk 5 above)
Comment (darkk) [inline ntp.go]: Do you mean literally `QueryWithOptions` instead of `QueryEx`?

Actually, I've not found any good examples of naming convention for this case in golang stdlib to make it idiomatic, that's why I reused some ancient WinAPI convention.

File: ntp.go (hunk 6 above)
Comment (darkk) [inline ntp.go]: Yep, I'm sure. type `Conn` is just a wrapper for existing file descriptor as far as I could understand from the source, [example](https://godoc.org/golang.org/x/net/ipv4#example-Conn--MarkingTCP) also makes me think so.

File: ntp.go (hunk 7 above)
Comment (darkk) [inline ntp.go]: `m.OriginTime` becomes echo of `xmt` in incoming packet, but `m.TransmitTime` is overwritten with server timestamp of outgoing packet, so it's not equal `m.OriginTime`.

File: ntp.go (hunk 9 above)
Comment (darkk) [inline ntp.go]: By the way, what's the reason for `Time` to fallback to returning _local_ time in case of NTP client failure?

File: ntp.go (hunk 8 above)
Comment (darkk) [inline ntp.go]: There are some of them in [RFC5905](https://tools.ietf.org/html/rfc5905#section-8).

> A packet is bogus if the origin timestamp t1 in the packet does not match the xmt state variable T1.

**Comment by beevik** [comment]
Thanks for the update!

File: ntp.go (hunk 4 above)
Comment (stapelberg) [inline ntp.go]: FWIW, this indeed broke https://github.com/gokrazy/gokrazy/tree/master/cmd/ntp.
